sound/soc/sof/ipc4-pcm.c

```diff
@@ -421,13 +423,42 @@ static int sof_ipc4_pcm_trigger(struct snd_soc_component *component,
 		return -EINVAL;
 	}
 
+	pipeline_list = sof_ipc4_get_pipeline_list(component, substream);
+
+	/* nothing to trigger if the list is empty */
+	if (!pipeline_list || !pipeline_list->pipelines || !pipeline_list->count)
+		return 0;
+
+	/*
+	 * If use_chain_dma attribute is set we proceed to chained DMA
+	 * trigger function that handles the rest for the substream.
+	 */
+	if (sof_ipc4_is_chain_dma_in_use(pipeline_list))
+		return sof_ipc4_chain_dma_trigger(sdev, pipeline_list, state, cmd);
+
 	/* set the pipeline state */
 	return sof_ipc4_trigger_pipelines(component, substream, state, cmd);
 }
 
 static int sof_ipc4_pcm_hw_free(struct snd_soc_component *component,
 				struct snd_pcm_substream *substream)
 {
+	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(component);
+	struct snd_sof_pcm_stream_pipeline_list *pipeline_list;
+
+	pipeline_list = sof_ipc4_get_pipeline_list(component, substream);
+
+	/* nothing to trigger if the list is empty */
+	if (!pipeline_list || !pipeline_list->pipelines || !pipeline_list->count)
+		return 0;
+
+	/*
+	 * If use_chain_dma attribute is set we proceed to chained DMA
+	 * trigger function that handles the rest for the substream.
+	 */
+	if (sof_ipc4_is_chain_dma_in_use(pipeline_list))
+		return sof_ipc4_chain_dma_trigger(sdev, pipeline_list, SOF_IPC4_PIPE_RESET, 0);
+
 	/* command is not relevant with RESET, so just pass 0 */
 	return sof_ipc4_trigger_pipelines(component, substream, SOF_IPC4_PIPE_RESET, 0);
```

Comment:
If you would introduce:
```c
static int sof_ipc4_change_pipelines_state(struct snd_soc_component *component,
					   struct snd_pcm_substream *substream,
					   int state, int cmd)
{
	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(component);
	struct snd_sof_pcm_stream_pipeline_list *pipeline_list;

	pipeline_list = sof_ipc4_get_pipeline_list(component, substream);

	/* list is empty, nothing to do */
	if (!pipeline_list || !pipeline_list->pipelines || !pipeline_list->count)
		return 0;

	/* Use either ChainDMA or pipeline triggering function to set the new state */
	if (sof_ipc4_is_chain_dma_in_use(pipeline_list))
		return sof_ipc4_chain_dma_trigger(sdev, pipeline_list, state, cmd);

	return sof_ipc4_trigger_pipelines(component, substream, state, cmd);
}
```
then here you could call it:
```c
	/* command is not relevant with RESET, so just pass 0 */
	sof_ipc4_change_pipelines_state(component, substream, SOF_IPC4_PIPE_RESET, 0);
```
and use the same helper in the trigger function.
I think it would make the code cleaner and self explanatory in most places.

Reply:
and while at it, it might worth to change the `sof_ipc4_trigger_pipelines()` parameter list to `sdev, pipeline_list, state, cmd` also?